sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/UnsafeArrayDataBenchmark.scala

```diff
@@ -0,0 +1,251 @@
+import org.apache.spark.sql.catalyst.expressions.codegen.{BufferHolder, UnsafeArrayWriter}
+import org.apache.spark.util.Benchmark
+
+/**
+ * Benchmark [[UnsafeArrayDataBenchmark]] for UnsafeArrayData
+ * To run this:
+ *  1. replace ignore(...) with test(...)
+ *  2. build/sbt "sql/test-only *benchmark.UnsafeArrayDataBenchmark"
+ *
+ * Benchmarks in this file are skipped in normal builds.
+ */
+class UnsafeArrayDataBenchmark extends BenchmarkBase {
+
+  def calculateHeaderPortionInBytes(count: Int) : Int = {
+    // Use this assignment for SPARK-15962
+    // val size = 4 + 4 * count
+    val size = UnsafeArrayData.calculateHeaderPortionInBytes(count)
+    size
+  }
+
+  def readUnsafeArray(iters: Int): Unit = {
+    val count = 1024 * 1024 * 16
+    val rand = new Random(42)
+
+    var intResult: Int = 0
+    val intBuffer = Array.fill[Int](count) { rand.nextInt }
+    val intEncoder = ExpressionEncoder[Array[Int]].resolveAndBind()
+    val intInternalRow = intEncoder.toRow(intBuffer)
+    val intUnsafeArray = intInternalRow.getArray(0)
+    val readIntArray = { i: Int =>
+      var n = 0
+      while (n < iters) {
+        val len = intUnsafeArray.numElements
+        var sum = 0.toInt
+        var i = 0
+        while (i < len) {
+          sum += intUnsafeArray.getInt(i)
+          i += 1
+        }
+        intResult = sum
+        n += 1
+      }
+    }
+
+    var doubleResult: Double = 0
+    val doubleBuffer = Array.fill[Double](count) { rand.nextDouble }
+    val doubleEncoder = ExpressionEncoder[Array[Double]].resolveAndBind()
+    val doubleInternalRow = doubleEncoder.toRow(doubleBuffer)
+    val doubleUnsafeArray = doubleInternalRow.getArray(0)
+    val readDoubleArray = { i: Int =>
+      var n = 0
+      while (n < iters) {
+        val len = doubleUnsafeArray.numElements
+        var sum = 0.toDouble
+        var i = 0
+        while (i < len) {
+          sum += doubleUnsafeArray.getDouble(i)
+          i += 1
+        }
+        doubleResult = sum
+        n += 1
+      }
+    }
+
+    val benchmark = new Benchmark("Read UnsafeArrayData", count * iters)
+    benchmark.addCase("Int")(readIntArray)
+    benchmark.addCase("Double")(readDoubleArray)
+    benchmark.run
+    /*
+    Java HotSpot(TM) 64-Bit Server VM 1.8.0_92-b14 on Mac OS X 10.10.4
+    Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
+
+    Read UnsafeArrayData:                    Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+    ------------------------------------------------------------------------------------------------
+    Int                                            279 /  294        600.4           1.7       1.0X
+    Double                                         296 /  303        567.0           1.8       0.9X
+    */
+  }
+
+  def writeUnsafeArray(iters: Int): Unit = {
+    val count = 1024 * 1024 * 16
+
+    val intUnsafeRow = new UnsafeRow(1)
+    val intUnsafeArrayWriter = new UnsafeArrayWriter
```

Comment:
Got it. My interpretation was to use an `UnsafeArray` generated by `encoder.toRow(array)` for benchmark. I will update `writeUnsafeArray` to measure the elapsed time of `encoder.toRow(array)`